ambry-commons/src/main/java/com/github/ambry/commons/RetryExecutor.java

```diff
@@ -67,8 +67,15 @@ public RetryExecutor(ScheduledExecutorService scheduler) {
           int waitTimeMs = policy.waitTimeMs(currAttempts);
           logger.info("{} of {} attempts failed, will keep retrying after a {} ms backoff. exception='{}'",
               currAttempts, policy.maxAttempts(), waitTimeMs, exception);
-          scheduler.schedule(() -> recursiveAsyncRetry(call, isRetriable, policy, userCallback, currAttempts), waitTimeMs,
-              TimeUnit.MILLISECONDS);
+          if (scheduler == null) {
```

Comment:
I see the case where scheduler is null. Can we also pass in a scheduler for that case? To avoid the if else here?

Reply:
Discussed offline. Adding comments to explain each case clearly.